onnxruntime/test/onnx/testcase_driver.h

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "testcase_request.h"
+#include "testenv.h"
+
+#include "core/common/common.h"
+#include <mutex>
+#include <memory>
+#include <condition_variable>
+#include <vector>
+
+class TestCaseResult;
+
+namespace onnxruntime {
+namespace test {
+
+/// <summary>
+/// Drives test execution
+/// </summary>
+class TestCaseDriver {
+ public:
+  /// <summary>
+  /// Runs all the test cases sequentially. If there are several data tests
+  /// in each test case they would be run in parallel.
+  /// </summary>
+  /// <param name="env">test environment</param>
+  /// <param name="concurrent_runs">number of data tests within the model to run concurrently if,
+  /// if 1, data tests are ran sequentially.
+  /// </param>
+  /// <param name="repeat_count">Repeat each data tests this many times (only for non-concurrent execution)</param>
+  /// <returns>All tests results</returns>
+  static std::vector<std::shared_ptr<TestCaseResult>> Run(const TestEnv& env,
+                                                          size_t concurrent_runs, size_t repeat_count);
+
+  /// <summary>
+  /// Runs all test cases(models) concurrently but not more than
+  /// parallel_models
+  /// </summary>
+  /// <param name="env">test environment</param>
+  /// <param name="parallel_models">number of parallel models (test cases) to run concurrently</param>
+  /// <param name="concurrent_runs">number of data tests to run concurrently on a specific test case(model)</param>
+  /// <returns>All test results</returns>
+  static std::vector<std::shared_ptr<TestCaseResult>> RunParallel(const TestEnv& env, size_t parallel_models,
+                                                                  size_t concurrent_runs);
+
+  ORT_DISALLOW_ASSIGNMENT(TestCaseDriver);
+
+ private:
+  TestCaseDriver(const TestEnv& env, size_t concurrent_runs)
+      : env_(env),
+        concurrent_runs_(concurrent_runs) {
+    results_.reserve(env.GetTests().size());
+    CallableFactory<TestCaseDriver, void, std::shared_ptr<TestCaseResult>> f(this);
+    on_test_case_complete_ = f.GetCallable<&TestCaseDriver::OnTestCaseComplete>();
+  }
+
+  ~TestCaseDriver() = default;
```

Comment:
What does this do?

Reply:
This declares the destructor as private

---
In reply to: [488992820](https://github.com/microsoft/onnxruntime/pull/5169#discussion_r488992820) [](ancestors = 488992820)